src/main/java/com/aws/greengrass/deployment/ConnectivityValidator.java

```diff
@@ -0,0 +1,224 @@
+        return !currentNetworkProxy.equals(newNetworkProxy);
+    }
+
+    private boolean awsRegionHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        Topic currentAwsRegion = currentdeviceConfiguration.getAWSRegion();
+        Topic newAwsRegion = deviceConfiguration.getAWSRegion();
+        return !currentAwsRegion.equals(newAwsRegion);
+    }
+
+    private boolean mqttHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        Topics currentMqtt = currentdeviceConfiguration.getMQTTNamespace();
+        Topics newMqtt = deviceConfiguration.getMQTTNamespace();
+        return !currentMqtt.equals(newMqtt);
+    }
+
+    private boolean iotDataEndpointHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        Topic currentIotDataEndpoint = currentdeviceConfiguration.getIotDataEndpoint();
+        Topic newIotDataEndpoint = deviceConfiguration.getIotDataEndpoint();
+        return !currentIotDataEndpoint.equals(newIotDataEndpoint);
+    }
+
+    private boolean greengrassDataPlaneEndpointHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        Topic currentGreengrassDataPlaneEndpoint = currentdeviceConfiguration.getGGDataEndpoint();
+        Topic newGreengrassDataPlaneEndpoint = deviceConfiguration.getGGDataEndpoint();
+        return !currentGreengrassDataPlaneEndpoint.equals(newGreengrassDataPlaneEndpoint);
+    }
+
+    private boolean greengrassDataPlanePortHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        int currentGreengrassDataPlanePort = Coerce.toInt(currentdeviceConfiguration
+                .getGreengrassDataPlanePort().getOnce());
+        int newGreengrassDataPlanePort = Coerce.toInt(deviceConfiguration.getGreengrassDataPlanePort().getOnce());
+        return currentGreengrassDataPlanePort != newGreengrassDataPlanePort;
+    }
+
+    /**
+     * Creates an MQTT client and checks if the device can connect to AWS.
+     *
+     * @param totallyCompleteFuture Future will be updated if validation fails
+     * @throws InterruptedException if sleep is interrupted
+     */
+    public boolean mqttClientCanConnect(CompletableFuture<DeploymentResult> totallyCompleteFuture)
+            throws InterruptedException {
+        logger.atDebug().log("Checking MQTT client can connect");
+
+        builder = mqttClient.createMqttConnectionBuilder(deviceConfiguration, null);
+        String clientId = Coerce.toString(deviceConfiguration.getThingName()) + "#validation";
+        connection = builder.withClientId(clientId).build();
+
+        for (int i = 0; i < maxRetries; i++) {
+            try {
+                connection.connect()
+                        .get(mqttClient.getMqttOperationTimeoutMillis(deviceConfiguration), TimeUnit.MILLISECONDS);
+                return true;
+            } catch (MqttException | TimeoutException | ExecutionException e) {
+                Thread.sleep(TimeUnit.SECONDS.toMillis(2));
+            }
```

Comment:
is this exception message useful for the deployment result? should customer see this error?

Reply:
yes at least we should log them. If we can know which config value is wrong then we should put them in the deployment failure message.

Reply:
Currently I don't see a way to determine which config value is wrong based off the exception messages.